### DISCUSSION TITLE
SHA256 mismatch

### DISCUSSION THREAD

**Discussion by miguelb39** [Homebrew/discussions#588] [Everyday usage] [answered]
Hi guys,

Getting a SHA256 mismatch error on 'shutter-encoder' ... looks like this:

```
miguel@iMac3 ~ % brew install shutter-encoder
                              
==> Downloading https://www.shutterencoder.com/Shutter%20Encoder%20(MAC%20Version%2014.4).zip
==> Downloading from https://www.shutterencoder.com/Shutter%20Encoder%20(MAC%20Version%2014.6).zip
######################################################################## 100.0%
Error: SHA256 mismatch
Expected: 1e7107d399b9c40c30a8a16d8f283e4499a80c7f5ae35c4f5e4cae293e077be5
  Actual: 7504fa09c384609089b8817d5971195bc34f3f18d3dd4ef0796b2310b5cf606f
    File: /Users/miguel/Library/Caches/Homebrew/downloads/f04852a2581d7448b6230c16cceec031d7d15a365fb38f968c31ff22b9391fee--Shutter Encoder (MAC Version 14.6).zip
To retry an incomplete download, remove the file above.
```

I'm sure it's just a newer version or something, but I don't know how to proceed.

I'm on 10.15.7, and homebrew is up to date.

**Comment by cesarcoatl** [comment, marked as answer]
That's weird, because the hash is correct for that version, and the cask formula is correct. But apparently the `url` from the cask downloads 14.6 because version 14.4 was moved to https://www.shutterencoder.com/old%20versions/Mac/Shutter%20Encoder%20(MAC%20Version%2014.4).zip, this is not an error on the cask or caused by Homebrew, but the Website.

I just opened a PR [Homebrew/homebrew-cask#98498](https://github.com/Homebrew/homebrew-cask/pull/98498), so once it's merged, just `brew update` and try (re)installing.

**Reply by cesarcoatl** [reply]
I guess it's not as simple as a `bump-cask-pr`. It needs `desc` and `zap`.

**Reply by cesarcoatl** [reply]
PR awaiting approval from a maintainer.

**Reply by miguelb39** [reply]
Thanks!  I figured it was something related to the redirected URL :-)

**Reply by cesarcoatl** [reply]
https://github.com/Homebrew/homebrew-cask/pull/98498 has been merged and ready for you to install it.

```bash
$ brew info shutter-encoder
shutter-encoder: 14.6
https://www.shutterencoder.com/
Not installed
From: https://github.com/Homebrew/homebrew-cask/blob/HEAD/Casks/shutter-encoder.rb
==> Name
Shutter Encoder
==> Description
Video, audio and image converter
==> Artifacts
Shutter Encoder.app (App)
==> Analytics
install: 5 (30 days), 8 (90 days), 25 (365 days)
```

**Comment by JakkuSakura** [comment]
Almost everything for me is SHA256 mismatch, and I'm sure the mirror is fine since I have changed from tuna mirror
```log
==> Downloading https://mirrors.ustc.edu.cn/homebrew-bottles/bottles/libyaml-0.2.5.arm64_big_sur.bottle.tar.gz
Already downloaded: /Users/jack/Library/Caches/Homebrew/downloads/8ef8bff851906b0536c391ce117c3d9f14041506e4b89d33171418ce334f888f--libyaml-0.2.5.arm64_big_sur.bottle.tar.gz
Error: SHA256 mismatch
Expected: fe1082f3475a144261b41e2c3e0728b9331911b1cbfadfbc1f3d70d454709154
  Actual: 1bf0828f67cc4ffdcf84eb95a7f51115132b181c6d4ccf145c3d1873ebdb5f3d
    File: /Users/jack/Library/Caches/Homebrew/downloads/8ef8bff851906b0536c391ce117c3d9f14041506e4b89d33171418ce334f888f--libyaml-0.2.5.arm64_big_sur.bottle.tar.gz
To retry an incomplete download, remove the file above.
Warning: Bottle installation failed: building from source.
==> Downloading https://github.com/yaml/libyaml/archive/0.2.5.tar.gz
Already downloaded: /Users/jack/Library/Caches/Homebrew/downloads/688fdcea5b88140cb83d1f72e3a77fa76b6560f0d66eabcea7d54cf9f06d5e72--libyaml-0.2.5.tar.gz
==> Downloading https://mirrors.ustc.edu.cn/homebrew-bottles/bottles/gdbm-1.18.1_1.arm64_big_sur.bottle.tar.gz
Already downloaded: /Users/jack/Library/Caches/Homebrew/downloads/77233b168e5be5024eaecc47df2eff4ddcc91000bf77922404bbbd96693ef860--gdbm-1.18.1_1.arm64_big_sur.bottle.tar.gz
Error: SHA256 mismatch
Expected: 801b2bf95118871ee206de507131325613a1aa59ab7809032bb456f1b5f01a89
  Actual: 090ec5647502d8a23d813f7e6aea328b77547f78b2bf69b76a817656ecb09e97
    File: /Users/jack/Library/Caches/Homebrew/downloads/77233b168e5be5024eaecc47df2eff4ddcc91000bf77922404bbbd96693ef860--gdbm-1.18.1_1.arm64_big_sur.bottle.tar.gz
To retry an incomplete download, remove the file above.
Warning: Bottle installation failed: building from source.
==> Downloading https://ftp.gnu.org/gnu/gdbm/gdbm-1.18.1.tar.gz
######################################################################## 100.0%
==> Downloading https://mirrors.ustc.edu.cn/homebrew-bottles/bottles/readline-8.1.arm64_big_sur.bottle.tar.gz
######################################################################## 100.0%
Error: SHA256 mismatch
Expected: 940e7c2b80ef7f59b26726a5669a31fcb8ba7cbbb17eb1f2ca589dafa6e68e5e
  Actual: 26971810ce31d63cd0966c733a132b47e7df5eb790dcbd3f38947cc5d3114ce0
    File: /Users/jack/Library/Caches/Homebrew/downloads/ad46a5a60b2948a91c235d8e24bb732d859473901d6b368f1810770ddd7048cb--readline-8.1.arm64_big_sur.bottle.tar.gz
To retry an incomplete download, remove the file above.
Warning: Bottle installation failed: building from source.
==> Downloading https://ftp.gnu.org/gnu/readline/readline-8.1.tar.gz
```

**Reply by gromgit** [reply]
> Almost everything for me is SHA256 mismatch, and I'm sure the mirror is fine
> ```
> ==> Downloading https://mirrors.ustc.edu.cn/homebrew-bottles/bottles/libyaml-0.2.5.arm64_big_sur.bottle.tar.gz
> Already downloaded: /Users/jack/Library/Caches/Homebrew/downloads/8ef8bff851906b0536c391ce117c3d9f14041506e4b89d33171418ce334f888f--libyaml-0.2.5.arm64_big_sur.bottle.tar.gz
> Error: SHA256 mismatch
> Expected: fe1082f3475a144261b41e2c3e0728b9331911b1cbfadfbc1f3d70d454709154
>   Actual: 1bf0828f67cc4ffdcf84eb95a7f51115132b181c6d4ccf145c3d1873ebdb5f3d
>     File: /Users/jack/Library/Caches/Homebrew/downloads/8ef8bff851906b0536c391ce117c3d9f14041506e4b89d33171418ce334f888f--libyaml-0.2.5.arm64_big_sur.bottle.tar.gz
> ```

The expected checksums are the official ones from the Homebrew core tap, so clearly the bottles you're downloading are _not_ the official Homebrew ones. Whoever maintains the mirror you're using likely rebuilt the bottles, so it's no longer Homebrew's problem if they don't work as advertised...or do additional stuff that you really _don't_ want to happen.

**Reply by jonchang** [reply]
This usually happens when the Homebrew maintainers delete and re-upload published bottles that were committed to a formula's `bottle` block. In this case, the old bottle was in homebrew/core for almost 2 weeks before it was swapped out, so plenty of time for the outdated bottle to be picked up by an artifact mirror, which has the (correct, outdated) checksum for the arm64 big sur bottle.

We (the maintainers) try to avoid doing this, since it means existing bottle downloads and artifact mirrors start to throw checksum errors, which is not great for anyone, but sometimes these things slip through the cracks.

**Reply by JakkuSakura** [reply]
Thanks for your explanation! What if you update the bottle with a slightly different name, like with a version number(or hash)? So that these errors can be gone.
Example:
old `https://www.xxx.com/file.zip` hash=abcdef
new `https://www.xxx.com/file.zip?version=2` hash=fecabd
If a mirror stays in old, it's fine without errors
